File: elixir/serviceradar_core/lib/serviceradar/event_writer/otel_id.ex
defmodule ServiceRadar.EventWriter.OtelId do
  @moduledoc """
  Canonical OTel identifier normalization.

  The canonical contract shared across the whole pipeline:

  - `trace_id` is a 32-character lowercase hex string
  - `span_id` / `parent_span_id` are 16-character lowercase hex strings
  - absent/invalid/zero identifiers are `nil` (stored as SQL NULL)

  Incoming values can arrive in several encodings, depending on producer:

  - raw OTLP bytes (16 bytes trace / 8 bytes span)
  - ASCII hex text in OTLP bytes fields (the Erlang OTLP logs exporter
    `otel_otlp_logs.erl` copies hex Logger metadata verbatim into the
    protobuf bytes fields)
  - already-hex JSON strings, possibly uppercase
  - legacy double-hex strings (hex encoding of the ASCII hex string),
    arriving via JSON re-publication of previously mis-encoded rows
  - standard base64 (common when protobuf bytes round-trip through JSON)

  Normalization is intentionally idempotent: feeding an already-canonical
  id back through any of these functions returns it unchanged, and hex
  text is never hex-encoded a second time.
  """

  @doc """
  Normalizes a trace id to 32-character lowercase hex, or `nil`.
  """
  @spec normalize_trace_id(term()) :: String.t() | nil
  def normalize_trace_id(value), do: normalize(value, 16)

  @doc """
  Normalizes a span id to 16-character lowercase hex, or `nil`.
  """
  @spec normalize_span_id(term()) :: String.t() | nil
  def normalize_span_id(value), do: normalize(value, 8)

  @doc """
  Normalizes a parent span id to 16-character lowercase hex, or `nil`.

  All-zero parents (the protobuf encoding of "no parent") and empty
  values normalize to `nil`, so root spans always store NULL.
  """
  @spec normalize_parent_span_id(term()) :: String.t() | nil
  def normalize_parent_span_id(value), do: normalize(value, 8)

  # byte_len is the raw OTLP byte length: 16 for trace ids, 8 for span ids.
  defp normalize(nil, _byte_len), do: nil
  defp normalize("", _byte_len), do: nil

  defp normalize(value, byte_len) when is_binary(value) do
    hex_len = byte_len * 2

    cond do
      # Already-hex text (CRITICAL: prevents double-hex encoding).
      byte_size(value) == hex_len and hex?(value) ->
        finalize(String.downcase(value))

      # Legacy double-hex: hex encoding of the ASCII hex string.
      byte_size(value) == hex_len * 2 and hex?(value) ->
        decode_double_hex(value, hex_len)

      # Raw OTLP bytes.
      byte_size(value) == byte_len ->
        finalize(Base.encode16(value, case: :lower))

      # Standard base64 of the raw bytes.
      true ->
        decode_base64(value, byte_len)
    end
  end

  defp normalize(_value, _byte_len), do: nil

  defp decode_double_hex(value, hex_len) do
    case Base.decode16(value, case: :mixed) do
      {:ok, decoded} when byte_size(decoded) == hex_len ->
        if hex?(decoded) do
          finalize(String.downcase(decoded))
        end

      _ ->
        nil
    end
  end

  defp decode_base64(value, byte_len) do
    case Base.decode64(value) do
      {:ok, decoded} when byte_size(decoded) == byte_len ->
        finalize(Base.encode16(decoded, case: :lower))

      _ ->
        nil
    end
  end

  defp finalize(hex) do
    if all_zero?(hex), do: nil, else: hex
  end

  defp hex?(value) do
    for(<<char <- value>>, reduce: true) do
      acc -> acc and hex_char?(char)
    end
  end

  defp hex_char?(char) when char in ?0..?9 when char in ?a..?f when char in ?A..?F, do: true

  defp hex_char?(_char), do: false

  defp all_zero?(hex) do
    for(<<char <- hex>>, reduce: true) do
      acc -> acc and char == ?0
    end
  end
end

File: go/pkg/consumers/db-event-writer/otelid.go
package dbeventwriter

// Canonical OTel ID normalization (refactor-otel-signal-correlation, D1).
//
// Contract: trace_id = 32-char lowercase hex, span_id/parent_span_id =
// 16-char lowercase hex. Absent, invalid, or all-zero ids normalize to the
// empty string; the database layer stores empty ids as SQL NULL.
//
// Incoming ids arrive in several shapes depending on the producer:
//   - raw protobuf bytes (N bytes: 16 trace / 8 span)
//   - ASCII hex text inside protobuf bytes fields (the Erlang OTLP exporter
//     copies hex logger metadata verbatim into LogRecord trace_id/span_id)
//   - legacy double-hex text (hex of the ASCII hex string, via JSON)
//   - standard base64 (protobuf JSON encoding of bytes fields)

import (
	"encoding/base64"
	"encoding/hex"
	"strings"
)

const (
	traceIDByteLen = 16
	spanIDByteLen  = 8
)

// NormalizeTraceID canonicalizes a trace id from raw bytes or text into
// 32-char lowercase hex. Absent/invalid/all-zero inputs yield "".
func NormalizeTraceID[T ~string | ~[]byte](id T) string {
	return normalizeOTELID(string(id), traceIDByteLen)
}

// NormalizeSpanID canonicalizes a span id from raw bytes or text into
// 16-char lowercase hex. Absent/invalid/all-zero inputs yield "".
func NormalizeSpanID[T ~string | ~[]byte](id T) string {
	return normalizeOTELID(string(id), spanIDByteLen)
}

// NormalizeParentSpanID canonicalizes a parent span id. OTLP root spans may
// carry 8 zero bytes (or zero hex) for the parent; those normalize to "".
func NormalizeParentSpanID[T ~string | ~[]byte](id T) string {
	return normalizeOTELID(string(id), spanIDByteLen)
}

// normalizeOTELID applies the canonical id contract for an id of expected
// raw byte length byteLen. The branches are ordered: raw bytes, ASCII hex,
// legacy double-hex, base64. Anything else is rejected to "".
func normalizeOTELID(raw string, byteLen int) string {
	if raw == "" {
		return ""
	}

	hexLen := byteLen * 2
	canonical := ""

	switch {
	case len(raw) == byteLen:
		// Raw protobuf bytes.
		canonical = hex.EncodeToString([]byte(raw))
	case len(raw) == hexLen && isASCIIHex(raw):
		// Already canonical hex (possibly uppercase). Lowering instead of
		// re-encoding prevents double-encoding ASCII-hex-in-bytes payloads.
		canonical = strings.ToLower(raw)
	case len(raw) == hexLen*2 && isASCIIHex(raw):
		// Legacy double-hex: hex of the ASCII hex string. Decode once and
		// require the decoded payload to itself be ASCII hex.
		if decoded, err := hex.DecodeString(raw); err == nil {
			if inner := string(decoded); isASCIIHex(inner) {
				canonical = strings.ToLower(inner)
			}
		}
	}

	if canonical == "" {
		if decoded, err := base64.StdEncoding.DecodeString(raw); err == nil && len(decoded) == byteLen {
			canonical = hex.EncodeToString(decoded)
		}
	}

	if canonical == "" {
		otelIDRejected.Add(1)
		return ""
	}

	if isAllZeroHex(canonical) {
		return ""
	}

	return canonical
}

func isASCIIHex(s string) bool {
	if s == "" {
		return false
	}

	for i := 0; i < len(s); i++ {
		c := s[i]
		if (c < '0' || c > '9') && (c < 'a' || c > 'f') && (c < 'A' || c > 'F') {
			return false
		}
	}

	return true
}

func isAllZeroHex(s string) bool {
	for i := 0; i < len(s); i++ {
		if s[i] != '0' {
			return false
		}
	}

	return true
}
